### PR TITLE
DEVPROD-10420: do not link patch_optional dependencies for alias versions

### PR DESCRIPTION
**Pull request by Kimchelly**
DEVPROD-10420

### Description
The tasks that have hit the task dependencies alert are tasks that have a dependency on a nonexistent task. They get stuck in the task queue because they're waiting to run on the nonexistent task.

This happens because of `patch_optional` dependencies in non-commit waterfall versions. For example, let's say there's a `taskA` and it has a `patch_optional` dependency on `taskB`. The expected behavior of `patch_optional` is that the dependency will never be created automatically in the version, so the dependency link should only be created if both `taskA` and `taskB` are created together. Furthermore, let's assume there's a periodic build that has an alias that selects only `taskA` to run. In this case, `taskA` should be created and `taskB` shouldn't (and we do see this happen, which is correct). However, even though `taskB` wasn't created, `taskA` nonetheless has a `depends_on` link to the nonexistent `taskB` (this shouldn't happen).

The cause of the bug is that, when creating a waterfall version, we have to make a pre-generated table of task IDs. Currently, the pre-generated task ID table creates task IDs for _all_ possible variant-tasks, even those that might not be created upon version creation. [This logic](https://github.com/evergreen-ci/evergreen/blob/e688de4801991d8740f733c8256c4bb4175c84af/model/dependency.go#L216-L217) will skip the `patch_optional` dependency so the dependent task is not automatically created (correct), _but_ [this logic](https://github.com/evergreen-ci/evergreen/blob/e688de4801991d8740f733c8256c4bb4175c84af/model/lifecycle.go#L934) will still create the dependency link to the (nonexistent) task because the pre-generated task ID exists (not correct). To fix this, I filtered tasks from the task ID table so that for versions like periodic builds that create a subset of all possible tasks, they only generate task IDs for tasks that will actually be created.

* Do not create task IDs for tasks that won't exist.
    * Add safety check to error if somehow we try to create a task that has no task ID.
* Improve "Task dependencies" alert log to include the error.

### Testing
* Added unit tests. Verified that this unit test fails before the change, and passes afterwards.
* Ran a [periodic build](https://spruce-staging.corp.mongodb.com/version/66f5c0c9086856000733f38d/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC) in the sandbox to reproduce the issue of the task with a nonexistent dependency. Then applied my fix and verified that [the next periodic build](https://spruce-staging.corp.mongodb.com/version/66f5c45120eb3f00076c0118/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC) was fixed and the `patch_optional` dependency was ignored.

### Documentation
Clarified in documentation that, despite the name of the field, `patch_optional` means the dependency is optional in _all_ versions, not just patches. The docs already say this explicitly, but I thought it was worth clarifying that the name of the field doesn't correspond to what you'd expect it to do.

### CODE REVIEW

File: repotracker/repotracker_test.go [hunk 1]
```diff
@@ -1245,6 +1247,194 @@ tasks:
...
+		Alias:    alias.Alias,
+	}
+	v, err := CreateVersionFromConfig(s.ctx, projectInfo, metadata, false, nil)
+	s.NoError(err)
+	s.Require().NotNil(v)
+	s.Len(v.Errors, 0)
+
+	tasks, err := task.Find(task.ByVersion(v.Id))
+	s.NoError(err)
+	s.Require().Len(tasks, 2)
```
Comment (hadjri) [inline repotracker/repotracker_test.go:1413]: nit: don't need require here since we don't index the slice

File: model/lifecycle.go [hunk 2]
```diff
@@ -742,6 +742,9 @@ func createTasksForBuild(ctx context.Context, creationInfo TaskCreationInfo) (ta
 	taskMap := make(map[string]*task.Task)
 	for _, t := range tasksToCreate {
 		id := execTable.GetId(creationInfo.Build.BuildVariant, t.Name)
+		if id == "" {
```
Comment (hadjri) [inline model/lifecycle.go:745]: nit: can we move this into `createOneTask` for empty id protection against potential future uses of it?

**Review by hadjri — APPROVED** [review]
Good find!

File: repotracker/repotracker_test.go (hunk 1 above)
Comment (Kimchelly) [reply repotracker/repotracker_test.go]: Switched to assert.

**Pull request merged**